python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,223 @@
+
+    Return:
+        list of IRModule
+    """
+    mods = {}
+    string_config = [{}] * len(config)
+    for ir_mod in config:
+        mod_config = config[ir_mod]
+        string_config[mod_config["pipeline"]["mod_indx"] - 1] = mod_config["pipeline"]
+        build_func = relay.build
+        # if there is a self defined build function then use it.
+        if mod_config["build"]:
+            build_func = mod_config.build
+
+        mod = build_func(
+            ir_mod,
+            mod_config["target"],
+            params=mod_config["params"],
+            target_host=mod_config["target_host"],
+            mod_name=mod_config["mod_name"],
+        )
+
+        mods[mod] = {"dev": mod_config["dev"]}
+
+    return mods, string_config
+
+
+def create(mods, mod_config):
+    """Create a pipeline runtime executor.
+
+    Parameters
+    ----------
+    sub_mods :
+        {"lib": <module>,
+         "dev": <device>}
+
+    Returns
+    -------
+    submodule : PipelineModule
+        Runtime pipeline module.
+    """
+    pipeline_mods, string_config = build_pipeline(mod_config)
+
+    mods = []
+    for pipeline_mod in pipeline_mods:
+        mod = graph_executor.GraphModule(
+            pipeline_mod["default"](pipeline_mods[pipeline_mod]["dev"])
+        )
+
+        mods.append(mod)
+
+    submodule = PipelineModule(mods, json.dumps(string_config))
+    return submodule
+
+
+class PipelineModule(object):
+    """Wrapper runtime module.
+
+    This is a thin wrapper of the underlying TVM module.
+    you can also directly call set_input, run, and get_output
+    of underlying module functions
+
+    Parameters
+    ----------
+    module : tvm.runtime.Module
+        The internal tvm module that holds the actual graph functions.
+
+    Attributes
+    ----------
+    module : tvm.runtime.Module
+        The internal tvm module that holds the actual graph functions.
+
+    """
+
+    def __init__(self, graph_modules, pipeline_config):
+        mods = []
+        for module in graph_modules:
+            mods.append(module.module)
+
+        pipelinecreate = tvm._ffi.get_global_func("tvm.pipeline_executor.create")
+        module = pipelinecreate(mods, pipeline_config)
+
+        self.graph_modules_ = graph_modules
+
+        self._set_input = module["set_input"]
+        self._run = module["run"]
+        self._stop = module["stop"]
+        self._get_output = module["get_output"]
+        self._get_input = module["get_input"]
+        self._get_num_outputs = module["get_num_outputs"]
+        self._get_num_inputs = module["get_num_inputs"]
+
+    def set_input(self, key=None, value=None, params=None):
+        """Set inputs to the module via kwargs
+
+        Parameters
+        ----------
+        key : int or str
+           The input key
+
+        value : the input value.
+           The input key
+
+        params : dict of str to NDArray
+           Additional arguments
+        """
+        if key is not None:
```

Comment:
Why `key` and `value` are allowed to be None?

Reply:
fixed.